src/Native/Unix/System.Security.Cryptography.Native/pal_x509.c

```diff
@@ -323,3 +387,659 @@ X509* CryptoNative_X509UpRef(X509* x509)
+    return t;
+}
+
+X509VerifyStatusCode CryptoNative_X509ChainGetCachedOcspStatus(X509_STORE_CTX* storeCtx, char* cachePath)
+{
+    if (storeCtx == NULL || cachePath == NULL)
+    {
+        return (X509VerifyStatusCode)-1;
+    }
+
+    X509* subject;
+    X509* issuer;
+
+    if (!Get0CertAndIssuer(storeCtx, &subject, &issuer))
+    {
+        return (X509VerifyStatusCode)-2;
+    }
+
+    X509VerifyStatusCode ret = PAL_X509_V_ERR_UNABLE_TO_GET_CRL;
+    char* fullPath = BuildOcspCacheFilename(cachePath, subject);
+
+    if (fullPath == NULL)
+    {
+        return ret;
+    }
+
+    BIO* bio = BIO_new_file(fullPath, "rb");
+    OCSP_RESPONSE* resp = NULL;
+
+    if (bio != NULL)
+    {
+        resp = d2i_OCSP_RESPONSE_bio(bio, NULL);
+        BIO_free(bio);
+    }
+
+    if (resp != NULL)
+    {
+        ASN1_GENERALIZEDTIME* thisUpdate = NULL;
+        ASN1_GENERALIZEDTIME* nextUpdate = NULL;
+        ret = CheckOcsp(NULL, resp, subject, issuer, storeCtx, &thisUpdate, &nextUpdate);
+
+        if (ret != PAL_X509_V_ERR_UNABLE_TO_GET_CRL)
+        {
+            time_t oldest = GetIssuanceWindowStart();
+
+            // If either the thisUpdate or nextUpdate is missing we can't determine policy, so reject it.
+            // oldest = now - window;
+            //
+            // if thisUpdate < oldest || nextUpdate < now, reject.
+            //
+            // Since X509_cmp(_current)_time returns 0 on error, do a <= 0 check.
+            if (nextUpdate == NULL ||
+                thisUpdate == NULL ||
+                X509_cmp_current_time(nextUpdate) <= 0 ||
+                X509_cmp_time(thisUpdate, &oldest) <= 0)
+            {
+                ret = PAL_X509_V_ERR_UNABLE_TO_GET_CRL;
+            }
+        }
+
+        if (nextUpdate != NULL)
+        {
+            ASN1_GENERALIZEDTIME_free(nextUpdate);
+        }
+
+        if (thisUpdate != NULL)
+        {
+            ASN1_GENERALIZEDTIME_free(thisUpdate);
+        }
+    }
+
+    if (ret == PAL_X509_V_ERR_UNABLE_TO_GET_CRL)
+    {
+        unlink(fullPath);
+        ERR_clear_error();
+    }
+
+    free(fullPath);
+
+    if (resp != NULL)
+    {
+        OCSP_RESPONSE_free(resp);
+    }
+
+    return ret;
+}
+
+OCSP_REQUEST* CryptoNative_X509ChainBuildOcspRequest(X509_STORE_CTX* storeCtx)
+{
+    if (storeCtx == NULL)
+    {
+        return NULL;
+    }
+
+    X509* subject;
+    X509* issuer;
+
+    if (!Get0CertAndIssuer(storeCtx, &subject, &issuer))
+    {
+        return NULL;
+    }
+
+    OCSP_CERTID* certId = MakeCertId(subject, issuer);
+
+    if (certId == NULL)
+    {
+        return NULL;
+    }
+
+    OCSP_REQUEST* req = OCSP_REQUEST_new();
+
+    if (req == NULL)
+    {
+        OCSP_CERTID_free(certId);
+        return NULL;
+    }
+
+    if (!OCSP_request_add0_id(req, certId))
+    {
+        OCSP_CERTID_free(certId);
+        OCSP_REQUEST_free(req);
+        return NULL;
+    }
+
+    // Add a random nonce.
+    OCSP_request_add1_nonce(req, NULL, -1);
```

Comment:
Who frees the cert id here? Is that now part of req?

Reply:
Yeah.  `add0` means that the reference count was not incremented; and as an ASN.1-based type it recursively frees all of its children.

Since I just spent like 30 minutes quadruple checking that, I'll leave a note.